Unknown reference for resolved name `Self` in impl of `Clone` for `!`

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
// ---- gccrs additions

#[lang = "sized"]
trait Sized {}

// ---- gccrs additions

#[lang = "clone"]
pub trait Clone: Sized {
    #[stable(feature = "rust1", since = "1.0.0")]
    #[must_use = "cloning is often expensive and is not expected to have side effects"]
    fn clone(&self) -> Self;

    #[inline]
    #[stable(feature = "rust1", since = "1.0.0")]
    fn clone_from(&mut self, source: &Self) {
        *self = source.clone()
    }
}

#[unstable(feature = "never_type", issue = "35121")]
impl Clone for ! {
    #[inline]
    fn clone(&self) -> Self {
        *self
    }
}
```

I expected to see this happen: no errors

Instead, this happened:

```rust
test.rs:24:24: error: unknown reference for resolved name: ‘Self’
   24 |     fn clone(&self) -> Self {
      |                        ^~~~
crab1: internal compiler error: Segmentation fault
0x490e16c crash_signal(int)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/toplev.cc:319
0x23c1eec std::vector<std::pair<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::size() const
	/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1/../../../../include/c++/13.2.1/bits/stl_vector.h:990
0x23c1df0 std::vector<std::pair<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::vector(std::vector<std::pair<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&)
	/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1/../../../../include/c++/13.2.1/bits/stl_vector.h:599
0x23c1dac Rust::Resolver::CanonicalPath::CanonicalPath(Rust::Resolver::CanonicalPath const&)
	../../gcc/rust/util/rust-canonical-path.h:49
0x25f73f3 Rust::RustIdent::RustIdent(Rust::Resolver::CanonicalPath const&, unsigned int)
	../../gcc/rust/util/rust-identifier.h:33
0x26a302e Rust::Resolver::TypeCheckImplItem::visit(Rust::HIR::Function&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-implitem.cc:467
0x24d953c Rust::HIR::Function::accept_vis(Rust::HIR::HIRImplVisitor&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/hir/tree/rust-hir.cc:5011
0x26a25d8 Rust::Resolver::TypeCheckImplItem::Resolve(Rust::HIR::ImplBlock*, Rust::HIR::ImplItem*, Rust::TyTy::BaseType*, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-implitem.cc:337
0x2674394 Rust::Resolver::TypeCheckItem::resolve_impl_item(Rust::HIR::ImplBlock&, Rust::HIR::ImplItem&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-item.cc:497
0x26741da Rust::Resolver::TypeCheckItem::ResolveImplItem(Rust::HIR::ImplBlock&, Rust::HIR::ImplItem&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-type-check-item.cc:63
0x26de8c9 Rust::Resolver::query_type(unsigned int, Rust::TyTy::BaseType**)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-type-util.cc:82
0x26d613d Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)::$_1::operator()(unsigned int, Rust::HIR::ImplBlock*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-dot-operator.cc:228
0x26d5f09 bool std::__invoke_impl<bool, Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)::$_1&, unsigned int, Rust::HIR::ImplBlock*>(std::__invoke_other, Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)::$_1&, unsigned int&&, Rust::HIR::ImplBlock*&&)
	/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1/../../../../include/c++/13.2.1/bits/invoke.h:61
0x26d5eb4 std::enable_if<is_invocable_r_v<bool, Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)::$_1&, unsigned int, Rust::HIR::ImplBlock*>, bool>::type std::__invoke_r<bool, Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)::$_1&, unsigned int, Rust::HIR::ImplBlock*>(Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)::$_1&, unsigned int&&, Rust::HIR::ImplBlock*&&)
	/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1/../../../../include/c++/13.2.1/bits/invoke.h:114
0x26d5d9c std::_Function_handler<bool (unsigned int, Rust::HIR::ImplBlock*), Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)::$_1>::_M_invoke(std::_Any_data const&, unsigned int&&, Rust::HIR::ImplBlock*&&)
	/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1/../../../../include/c++/13.2.1/bits/std_function.h:290
0x2516576 std::function<bool (unsigned int, Rust::HIR::ImplBlock*)>::operator()(unsigned int, Rust::HIR::ImplBlock*) const
	/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1/../../../../include/c++/13.2.1/bits/std_function.h:591
0x250edf1 Rust::Analysis::Mappings::iterate_impl_blocks(std::function<bool (unsigned int, Rust::HIR::ImplBlock*)>)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/util/rust-hir-map.cc:826
0x26d3449 Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-hir-dot-operator.cc:203
0x26f1844 Rust::Resolver::AutoderefCycle::try_autoderefed(Rust::TyTy::BaseType*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-autoderef.cc:423
0x26f137c Rust::Resolver::AutoderefCycle::cycle(Rust::TyTy::BaseType*)
	/home/arthur/Git/rust-gcc/gccrs/build/gcc/../../gcc/rust/typecheck/rust-autoderef.cc:329
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
```

--- a/gcc/rust/resolve/rust-ast-resolve-type.cc
+++ b/gcc/rust/resolve/rust-ast-resolve-type.cc
@@ -69,7 +69,7 @@ ResolveType::visit (AST::InferredType &)
 void
 ResolveType::visit (AST::NeverType &)
 {
-  // FIXME
+  resolved_node = resolver->get_never_type_node_id ();
 }
 
 void
@@ -499,6 +499,12 @@ ResolveTypeToCanonicalPath::visit (AST::TraitObjectType &)
 {
   // FIXME is this actually allowed? dyn A+B
   rust_unreachable ();
+}
+
+void
+ResolveTypeToCanonicalPath::visit (AST::NeverType &type)
+{
+  result = CanonicalPath::new_seg (type.get_node_id (), "!");
 }
 
 ResolveTypeToCanonicalPath::ResolveTypeToCanonicalPath ()

--- a/gcc/rust/resolve/rust-ast-resolve-type.h
+++ b/gcc/rust/resolve/rust-ast-resolve-type.h
@@ -246,6 +246,8 @@ public:
 
   void visit (AST::TraitObjectType &type) override;
 
+  void visit (AST::NeverType &type) override;
+
 private:
   ResolveTypeToCanonicalPath ();
 

--- a/gcc/rust/resolve/rust-name-resolver.cc
+++ b/gcc/rust/resolve/rust-name-resolver.cc
@@ -429,7 +429,10 @@ Resolver::generate_builtins ()
   setup_builtin ("isize", isize);
   setup_builtin ("char", char_tyty);
   setup_builtin ("str", str);
-  setup_builtin ("!", never);
+
+  // never type
+  NodeId never_node_id = setup_builtin ("!", never);
+  set_never_type_node_id (never_node_id);
 
   // unit type ()
   TyTy::TupleType *unit_tyty
@@ -443,7 +446,7 @@ Resolver::generate_builtins ()
   set_unit_type_node_id (unit_type->get_node_id ());
 }
 
-void
+NodeId
 Resolver::setup_builtin (const std::string &name, TyTy::BaseType *tyty)
 {
   AST::PathIdentSegment seg (name, BUILTINS_LOCATION);
@@ -459,6 +462,8 @@ Resolver::setup_builtin (const std::string &name, TyTy::BaseType *tyty)
   mappings.insert_canonical_path (
     builtin_type->get_node_id (),
     CanonicalPath::new_seg (builtin_type->get_node_id (), name));
+
+  return builtin_type->get_node_id ();
 }
 
 void

--- a/gcc/rust/resolve/rust-name-resolver.h
+++ b/gcc/rust/resolve/rust-name-resolver.h
@@ -163,8 +163,12 @@ public:
   Scope &get_macro_scope () { return macro_scope; }
 
   NodeId get_global_type_node_id () { return global_type_node_id; }
+
   void set_unit_type_node_id (NodeId id) { unit_ty_node_id = id; }
   NodeId get_unit_type_node_id () { return unit_ty_node_id; }
+
+  void set_never_type_node_id (NodeId id) { never_ty_node_id = id; }
+  NodeId get_never_type_node_id () { return never_ty_node_id; }
 
   void push_new_module_scope (NodeId module_id)
   {
@@ -208,7 +212,7 @@ private:
   Resolver ();
 
   void generate_builtins ();
-  void setup_builtin (const std::string &name, TyTy::BaseType *tyty);
+  NodeId setup_builtin (const std::string &name, TyTy::BaseType *tyty);
 
   Analysis::Mappings &mappings;
   TypeCheckContext *tyctx;
@@ -222,6 +226,7 @@ private:
 
   NodeId global_type_node_id;
   NodeId unit_ty_node_id;
+  NodeId never_ty_node_id;
 
   // map a AST Node to a Rib
   std::map<NodeId, Rib *> name_ribs;

--- a/gcc/testsuite/rust/compile/issue-3035.rs
+++ b/gcc/testsuite/rust/compile/issue-3035.rs
@@ -1,0 +1,25 @@
+#[lang = "sized"]
+trait Sized {}
+
+// ---- gccrs additions
+
+#[lang = "clone"]
+pub trait Clone: Sized {
+    #[stable(feature = "rust1", since = "1.0.0")]
+    #[must_use = "cloning is often expensive and is not expected to have side effects"]
+    fn clone(&self) -> Self;
+
+    #[inline]
+    #[stable(feature = "rust1", since = "1.0.0")]
+    fn clone_from(&mut self, source: &Self) {
+        *self = source.clone()
+    }
+}
+
+#[unstable(feature = "never_type", issue = "35121")]
+impl Clone for ! {
+    #[inline]
+    fn clone(&self) -> Self {
+        *self
+    }
+}

--- a/gcc/testsuite/rust/compile/nr2/exclude
+++ b/gcc/testsuite/rust/compile/nr2/exclude
@@ -292,3 +292,4 @@ box_syntax_feature_gate.rs
 dropck_eyepatch_feature_gate.rs
 inline_asm_parse_output_operand.rs
 issue-3030.rs
+issue-3035.rs

